### PR TITLE
Use ModelSchema in adapter/serializer methods to ember-data

### PR DESCRIPTION
**Pull request by theroncross**
Ember Data passes a private
[ShimModelClass](https://github.com/emberjs/data/blob/3aa0fbddb7a650361bed44350c3b51669ec9065e/packages/store/addon/-private/legacy-model-support/shim-model-class.ts#L38) as the `type/modelClass/primaryModelClass` argument to many of the public methods, as opposed to an instance of the model, or even the class itself. The fallout is that the current types don't match the implementation, inclugin flagging any usage of `modelName`, a static property on the class.

This adds a `ModelSchema` interface, which matches the private, in-repo type implemented by `ShimModelClass`. It also updates the call sites with the new type, where appropiate.

Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `npm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

Select one of these and delete the others:

If adding a new definition:
- [ ] The package does not already provide its own types, or cannot have its `.d.ts` files generated via `--declaration`
- [ ] If this is for an npm package, match the name. If not, do not conflict with the name of an npm package.
- [ ] Create it with `dts-gen --dt`, not by basing it on an existing project.
- [ ] Represents shape of module/library [correctly](https://www.typescriptlang.org/docs/handbook/declaration-files/library-structures.html)
- [ ] `tslint.json` [should contain](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#linter-tslintjson) `{ "extends": "@definitelytyped/dtslint/dt.json" }`, and no additional rules.
- [ ] `tsconfig.json` [should have](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#tsconfigjson) `noImplicitAny`, `noImplicitThis`, `strictNullChecks`, and `strictFunctionTypes` set to `true`.

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: [ShimModelClass](https://github.com/emberjs/data/blob/3aa0fbddb7a650361bed44350c3b51669ec9065e/packages/store/addon/-private/legacy-model-support/shim-model-class.ts#L38)
- [ ] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.

If removing a declaration:
- [ ] If a package was never on Definitely Typed, you don't need to do anything. (If you wrote a package and provided types, you don't need to register it with us.)
- [ ] Delete the package's directory.
- [ ] Add it to `notNeededPackages.json`.

### CODE REVIEW

**Review by runspired — COMMENTED** [review]
Overall the change looks good, we'll need to get the tests passing.

Something we *could* do is make `ModelSchema` accept the generic such that the fields/attributes/relationships can be narrowed to those of a specific type's. I'm mixed on the actual utility of this in such a general context though and this is an area we intend to significantly change and refactor the types going forward anyway. It's likely safer to do this more generic thing.

**Comment by theroncross** [comment]
I'm still working on getting the glimmer types installed locally so I can run the tests. I _think_ the test changes are minimal enough I'll just let the tests run here and hope for the best.

I had considered making it generic, but didn't know how it would benefit. I suppose anything that accurately makes them more precise is a win.

File: package.json [hunk 1]
```diff
@@ -56,5 +56,8 @@
             "_comment": "This will remove husky from when we started migrating to use prettier",
             "pre-commit": "npm uninstall husky"
         }
+    },
+    "dependencies": {
```
Comment (peterblazejewicz) [inline package.json:60]: why this change? It's repo specific entry, not TS module specific one

Comment (theroncross) [reply package.json:60]: Sorry - added that for running tests. I'll drop it.

File: types/ember-data/index.d.ts [hunk 2]
```diff
@@ -1259,7 +1290,7 @@ export namespace DS {
          */
         findRecord<K extends keyof ModelRegistry>(
             store: Store,
-            type: ModelRegistry[K],
+            type: ModelSchema<K>,
```
[the remark this answers is not in the document]
Comment (theroncross) [reply types/ember-data/index.d.ts:1293]: AFAICT, and I believe @runspired confirmed, an instance of the ShimModelClass gets passed, not the model classes in the registry.

**Review by sheetalkamat — APPROVED** [review]
